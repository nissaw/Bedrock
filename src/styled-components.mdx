---
name: Why Styled Components?
route: /why-styled-components
---

# Why `styled-components`

The topic of CSS tooling is a controversial topic, at best. Pointless Twitter battles have been fought over why one should choose preprocessor like, Sass, Less, or PostCSS or CSS-in-JS
like styled-components or emotion. Others debate the need for these tools at all, debating that one should "just use the platform".

Ultimately, the debate is useless. Any and all these tools are useful and serve a specific purpose. So it begs the question, why styled-components?

Besides the [basic motivations](https://styled-components.com/docs/basics#motivation) that inspired the creation of the library itself, here are a few I find important:

## Zero Config

`style-components` just works. There are no build steps that you need to set up nor css files you need to add. Everything to do with the component is truly encapsulated, including the styles.
[Server side rendering](https://styled-components.com/docs/advanced#server-side-rendering) is also very easy to setup and well documented.

## Composable by default

Components already can be composed in the obvious way:

```javascript
const App = () => {
  return (
    <Stack gutter='md'>
      <PadBox padding='lg'>{/*  Content goes here */}</PadBox>
    </Stack>
  );
};
```

In addition to this, styled-components exposes an "`as`" prop that takes either a native element as a string, like "div" or "button" or a component that exposes the `className` prop.
This means we can rewrite the above like this:

```javascript
const App = () => {
  return (
    <Stack as={PadBox} gutter='md' padding='lg'>
      {/*  Content goes here */}
    </Stack>
  );
};
```

## Overriding default values with Theming

`styled-components` has a [powerful themeing capability](https://styled-components.com/docs/advanced#theming) that allows the default constants of the Bedrock Layout Components
to be overridden easily and universally in one location.

## Universally extensible via `className` and `style` props

`styled-components` exposes both the `className` prop and `style` prop, making them universally extensible no matter the system you are using to style your app.
