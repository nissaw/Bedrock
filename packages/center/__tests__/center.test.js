import React from "react";
import { create } from "react-test-renderer";
import { ThemeProvider } from "styled-components";
import Center from "../src";

const Lorem = () => (
  <>
    {Array.from(Array(4).keys()).map((i) => (
      <p key={i}>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nullam in
        vestibulum tortor, vitae venenatis lectus. Praesent gravida dapibus
        neque sit amet molestie. Morbi blandit eu dolor a luctus. Vestibulum
        sollicitudin elit ac nunc scelerisque rhoncus. Nulla felis sapien,
        condimentum ut imperdiet vel, aliquet id ante. Pellentesque habitant
        morbi tristique senectus et netus et malesuada fames ac turpis egestas.
        Quisque ultrices, quam nec scelerisque malesuada, lectus elit semper
        diam, ac placerat purus tortor et enim.
      </p>
    ))}
  </>
);

describe("Center", () => {
  describe("correct usage", () => {
    test("Center is not null", () => {
      expect(Center).toBeTruthy();
    });

    it("renders default width", () => {
      const center = create(
        <Center>
          <Lorem />
        </Center>
      );
      expect(center.toJSON()).toMatchSnapshot();
    });

    it("renders custom width", () => {
      const center = create(
        <Center maxWidth={320}>
          <Lorem />
        </Center>
      );
      expect(center.toJSON()).toMatchSnapshot();
    });

    it("renders with theme overrides", () => {
      const center = create(
        <ThemeProvider theme={{ breakPoints: { medium: 1600 } }}>
          <Center>
            <Lorem />
          </Center>
        </ThemeProvider>
      );
      expect(center.toJSON()).toMatchSnapshot();
    });
  });

  describe("incorrect usage", () => {
    let originalError;
    let spy;
    beforeEach(() => {
      originalError = console.error;
      spy = jest.fn();
      console.error = spy;
    });
    afterEach(() => {
      console.error = originalError;
    });

    it("renders default with console error with no children", () => {
      expect(spy.mock.calls.length).toBe(0);

      const errorStack = create(
        <Center maxWidth="incorrect">
          <Lorem />
        </Center>
      );

      expect(spy.mock.calls.length).toBe(1);
      expect(errorStack.toJSON()).toMatchSnapshot();
    });
  });
});
