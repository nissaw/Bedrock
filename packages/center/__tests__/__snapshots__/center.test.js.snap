// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Center correct usage renders custom width 1`] = `
.c0 {
  --maxWidth: 320px;
  box-sizing: content-box;
  margin-inline-start: auto;
  margin-inline-end: auto;
  margin-inline: auto;
  max-inline-size: var(--maxWidth);
}

<div
  className="c0"
>
  <p>
    Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nullam in vestibulum tortor, vitae venenatis lectus. Praesent gravida dapibus neque sit amet molestie. Morbi blandit eu dolor a luctus. Vestibulum sollicitudin elit ac nunc scelerisque rhoncus. Nulla felis sapien, condimentum ut imperdiet vel, aliquet id ante. Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas. Quisque ultrices, quam nec scelerisque malesuada, lectus elit semper diam, ac placerat purus tortor et enim.
  </p>
  <p>
    Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nullam in vestibulum tortor, vitae venenatis lectus. Praesent gravida dapibus neque sit amet molestie. Morbi blandit eu dolor a luctus. Vestibulum sollicitudin elit ac nunc scelerisque rhoncus. Nulla felis sapien, condimentum ut imperdiet vel, aliquet id ante. Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas. Quisque ultrices, quam nec scelerisque malesuada, lectus elit semper diam, ac placerat purus tortor et enim.
  </p>
  <p>
    Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nullam in vestibulum tortor, vitae venenatis lectus. Praesent gravida dapibus neque sit amet molestie. Morbi blandit eu dolor a luctus. Vestibulum sollicitudin elit ac nunc scelerisque rhoncus. Nulla felis sapien, condimentum ut imperdiet vel, aliquet id ante. Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas. Quisque ultrices, quam nec scelerisque malesuada, lectus elit semper diam, ac placerat purus tortor et enim.
  </p>
  <p>
    Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nullam in vestibulum tortor, vitae venenatis lectus. Praesent gravida dapibus neque sit amet molestie. Morbi blandit eu dolor a luctus. Vestibulum sollicitudin elit ac nunc scelerisque rhoncus. Nulla felis sapien, condimentum ut imperdiet vel, aliquet id ante. Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas. Quisque ultrices, quam nec scelerisque malesuada, lectus elit semper diam, ac placerat purus tortor et enim.
  </p>
</div>
`;

exports[`Center correct usage renders default width 1`] = `
.c0 {
  --maxWidth: 1023px;
  box-sizing: content-box;
  margin-inline-start: auto;
  margin-inline-end: auto;
  margin-inline: auto;
  max-inline-size: var(--maxWidth);
}

<div
  className="c0"
>
  <p>
    Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nullam in vestibulum tortor, vitae venenatis lectus. Praesent gravida dapibus neque sit amet molestie. Morbi blandit eu dolor a luctus. Vestibulum sollicitudin elit ac nunc scelerisque rhoncus. Nulla felis sapien, condimentum ut imperdiet vel, aliquet id ante. Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas. Quisque ultrices, quam nec scelerisque malesuada, lectus elit semper diam, ac placerat purus tortor et enim.
  </p>
  <p>
    Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nullam in vestibulum tortor, vitae venenatis lectus. Praesent gravida dapibus neque sit amet molestie. Morbi blandit eu dolor a luctus. Vestibulum sollicitudin elit ac nunc scelerisque rhoncus. Nulla felis sapien, condimentum ut imperdiet vel, aliquet id ante. Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas. Quisque ultrices, quam nec scelerisque malesuada, lectus elit semper diam, ac placerat purus tortor et enim.
  </p>
  <p>
    Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nullam in vestibulum tortor, vitae venenatis lectus. Praesent gravida dapibus neque sit amet molestie. Morbi blandit eu dolor a luctus. Vestibulum sollicitudin elit ac nunc scelerisque rhoncus. Nulla felis sapien, condimentum ut imperdiet vel, aliquet id ante. Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas. Quisque ultrices, quam nec scelerisque malesuada, lectus elit semper diam, ac placerat purus tortor et enim.
  </p>
  <p>
    Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nullam in vestibulum tortor, vitae venenatis lectus. Praesent gravida dapibus neque sit amet molestie. Morbi blandit eu dolor a luctus. Vestibulum sollicitudin elit ac nunc scelerisque rhoncus. Nulla felis sapien, condimentum ut imperdiet vel, aliquet id ante. Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas. Quisque ultrices, quam nec scelerisque malesuada, lectus elit semper diam, ac placerat purus tortor et enim.
  </p>
</div>
`;

exports[`Center correct usage renders with theme overrides 1`] = `
.c0 {
  --maxWidth: 1600px;
  box-sizing: content-box;
  margin-inline-start: auto;
  margin-inline-end: auto;
  margin-inline: auto;
  max-inline-size: var(--maxWidth);
}

<div
  className="c0"
>
  <p>
    Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nullam in vestibulum tortor, vitae venenatis lectus. Praesent gravida dapibus neque sit amet molestie. Morbi blandit eu dolor a luctus. Vestibulum sollicitudin elit ac nunc scelerisque rhoncus. Nulla felis sapien, condimentum ut imperdiet vel, aliquet id ante. Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas. Quisque ultrices, quam nec scelerisque malesuada, lectus elit semper diam, ac placerat purus tortor et enim.
  </p>
  <p>
    Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nullam in vestibulum tortor, vitae venenatis lectus. Praesent gravida dapibus neque sit amet molestie. Morbi blandit eu dolor a luctus. Vestibulum sollicitudin elit ac nunc scelerisque rhoncus. Nulla felis sapien, condimentum ut imperdiet vel, aliquet id ante. Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas. Quisque ultrices, quam nec scelerisque malesuada, lectus elit semper diam, ac placerat purus tortor et enim.
  </p>
  <p>
    Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nullam in vestibulum tortor, vitae venenatis lectus. Praesent gravida dapibus neque sit amet molestie. Morbi blandit eu dolor a luctus. Vestibulum sollicitudin elit ac nunc scelerisque rhoncus. Nulla felis sapien, condimentum ut imperdiet vel, aliquet id ante. Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas. Quisque ultrices, quam nec scelerisque malesuada, lectus elit semper diam, ac placerat purus tortor et enim.
  </p>
  <p>
    Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nullam in vestibulum tortor, vitae venenatis lectus. Praesent gravida dapibus neque sit amet molestie. Morbi blandit eu dolor a luctus. Vestibulum sollicitudin elit ac nunc scelerisque rhoncus. Nulla felis sapien, condimentum ut imperdiet vel, aliquet id ante. Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas. Quisque ultrices, quam nec scelerisque malesuada, lectus elit semper diam, ac placerat purus tortor et enim.
  </p>
</div>
`;

exports[`Center incorrect usage renders default with console error with no children 1`] = `
.c0 {
  --maxWidth: 1023px;
  box-sizing: content-box;
  margin-inline-start: auto;
  margin-inline-end: auto;
  margin-inline: auto;
  max-inline-size: var(--maxWidth);
}

<div
  className="c0"
>
  <p>
    Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nullam in vestibulum tortor, vitae venenatis lectus. Praesent gravida dapibus neque sit amet molestie. Morbi blandit eu dolor a luctus. Vestibulum sollicitudin elit ac nunc scelerisque rhoncus. Nulla felis sapien, condimentum ut imperdiet vel, aliquet id ante. Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas. Quisque ultrices, quam nec scelerisque malesuada, lectus elit semper diam, ac placerat purus tortor et enim.
  </p>
  <p>
    Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nullam in vestibulum tortor, vitae venenatis lectus. Praesent gravida dapibus neque sit amet molestie. Morbi blandit eu dolor a luctus. Vestibulum sollicitudin elit ac nunc scelerisque rhoncus. Nulla felis sapien, condimentum ut imperdiet vel, aliquet id ante. Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas. Quisque ultrices, quam nec scelerisque malesuada, lectus elit semper diam, ac placerat purus tortor et enim.
  </p>
  <p>
    Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nullam in vestibulum tortor, vitae venenatis lectus. Praesent gravida dapibus neque sit amet molestie. Morbi blandit eu dolor a luctus. Vestibulum sollicitudin elit ac nunc scelerisque rhoncus. Nulla felis sapien, condimentum ut imperdiet vel, aliquet id ante. Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas. Quisque ultrices, quam nec scelerisque malesuada, lectus elit semper diam, ac placerat purus tortor et enim.
  </p>
  <p>
    Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nullam in vestibulum tortor, vitae venenatis lectus. Praesent gravida dapibus neque sit amet molestie. Morbi blandit eu dolor a luctus. Vestibulum sollicitudin elit ac nunc scelerisque rhoncus. Nulla felis sapien, condimentum ut imperdiet vel, aliquet id ante. Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas. Quisque ultrices, quam nec scelerisque malesuada, lectus elit semper diam, ac placerat purus tortor et enim.
  </p>
</div>
`;
